src/Halogen/Aff/Driver/Eval.purs

```diff
@@ -115,12 +115,17 @@ eval lchs render r =
           handleLifecycle lchs (render ref)
           pure a
     Subscribe es next -> do
+      DriverState ({ subscriptions }) <- liftEff (readRef ref)
       forkAff do
         { producer, done } <- ES.unEventSource es
+        liftEff (modifyRef subscriptions (map (done : _)))
         let
           consumer = do
-            s <- lift <<< evalF ref =<< CR.await
-            when (s == ES.Listening) consumer
+            q <- CR.await
+            subs <- lift $ liftEff (readRef subscriptions)
+            when (isJust subs) do
+              s <- lift $ evalF ref q
+              when (s == ES.Listening) consumer
```

Comment:
Actually, just realised the `subscriptions` are only ever extended here, I should add something to remove them on process completion.